src/mscorlib/shared/System/MemoryExtensions.cs

```diff
@@ -771,17 +771,21 @@ public static bool EndsWith<T>(this ReadOnlySpan<T> span, ReadOnlySpan<T> value)
         /// </summary>
         public static void Reverse<T>(this Span<T> span)
         {
-            ref T p = ref MemoryMarshal.GetReference(span);
-            int i = 0;
-            int j = span.Length - 1;
-            while (i < j)
+            if (span.Length <= 1)
             {
-                T temp = Unsafe.Add(ref p, i);
-                Unsafe.Add(ref p, i) = Unsafe.Add(ref p, j);
-                Unsafe.Add(ref p, j) = temp;
-                i++;
-                j--;
+                return;
             }
+
+            ref T first = ref MemoryMarshal.GetReference(span);
+            ref T last = ref Unsafe.Add(ref Unsafe.Add(ref first, length), -1);
```

Comment:
Quick question to reviewers who are more familiar with the JIT and GC:

I played around with two different patterns here. The first, `Unsafe.Add(ref first, length - 1)`, generates this bytecode:

```asm
dec length
movsxd length64, length
lea last, [first + length * stride]
```

The second, `Unsafe.Add(ref Unsafe.Add(ref first, length), -1)`, generates this bytecode:

```asm
movsxd length64, length
lea last, [first + length * stride - stride]
```

Is it possible _in theory_ for the JIT to generate something like the below in the second case?

```asm
movsxd length64, length
lea last, [first + length * stride]
add last, -stride
```

And if it does generate this, is it possible that immediately after the `lea` instruction but before the `add`, _first_ and _last_ technically point to different objects, which could allow the GC to relocate _first_ but not _last_, causing the final `add` instruction to point to garbage?

Reply:
FYI, the reason I asked the above question is that this same pattern of "get a ref to just past the end of the array then back up" also appears in `Buffer.Memmove(ref byte, ref byte, ...)`. I don't recall that particular code pattern being mentioned in the review, so not sure if it just slipped under the radar during review or if it's actually safe.